node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
+                    assert_eq!(
+                        block.hash(),
+                        &fs.block_hash,
+                        "block loaded from storage should have a matching block hash."
                     )
                 }
-                Effects::new()
+                // Check if validator is bonded in the era in which the block was created in.
+                effect_builder
+                    .is_bonded_validator(fs.era_id, fs.public_key)
+                    .map(|is_bonded| {
+                        if is_bonded {
+                            Ok((maybe_block, fs, is_bonded))
+                        } else {
+                            // If it's not bonded in that era, we will check if it's bonded in the
+                            // future era.
+                            Err((maybe_block, fs))
+                        }
+                    })
             }
-            Event::GetBlockForFinalitySignaturesResult(_block_hash, Some(block)) => {
-                let old_count = block.proofs().len();
-                let (block, mut effects) =
-                    self.add_pending_finality_signatures(*block, effect_builder);
-                if block.proofs().len() > old_count {
-                    let block = Box::new(block);
-                    effects.extend(effect_builder.put_block_to_storage(block).ignore());
+            .result(
+                |(maybe_block, fs, is_bonded)| Event::IsBonded(maybe_block, fs, is_bonded),
+                |(maybe_block, fs)| Event::IsBondedFutureEra(maybe_block, fs),
+            ),
+            Event::IsBondedFutureEra(maybe_block, fs) => {
+                // This should be safe to do. `self.latest_block` will be None only if ew haven't
+                // finalized *any* block yet. which will be just after the Genesis
+                // era start. Even if we receive a finality signature before finalizing anything,
+                // that will be for the *current* era and we won't be looking up whether validator
+                // is bonded in the Contract Runtime.
+                let state_root_hash = self
+                    .latest_block
+                    .as_ref()
+                    .map(|lb| *lb.header().state_root_hash())
+                    .unwrap();
+                let next_era_id = self
+                    .latest_block
+                    .as_ref()
+                    .map(|lb| lb.header().era_id().successor())
+                    .unwrap();
+                // TODO: Use protocol version that is valid for the block's height.
+                let protocol_version = ProtocolVersion::new(SemVer::V1_0_0);
+                effect_builder
+                    .is_bonded_in_future_era(
+                        state_root_hash,
+                        next_era_id,
+                        protocol_version,
+                        fs.public_key,
+                    )
+                    .map(|is_bonded_res| match is_bonded_res {
```

Comment:
Can `result` be used for this instead of `map`?

Reply:
Yes, nice catch.

Reply:
Done in [`c0bd13f` (#683)](https://github.com/CasperLabs/casper-node/pull/683/commits/c0bd13f70e41208382667535631dd62c9a381c5c)